syntax = "proto3";

package secret_service.v1;
option go_package = "github.com/access-core-tech/watchflow/api/secret_service/v1";

import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";
import "google/protobuf/empty.proto";

service SecretStore {
  rpc InitStorage(InitStorageRequest) returns (InitStorageResponse);
  rpc Unseal(UnsealRequest) returns (google.protobuf.Empty);
  rpc Seal(SealRequest) returns (google.protobuf.Empty);
  rpc PutSecret(PutSecretRequest) returns (google.protobuf.Empty);
  rpc GetSecret(GetSecretRequest) returns (GetSecretResponse);
  rpc DeleteSecret(DeleteSecretRequest) returns (google.protobuf.Empty);
}

message InitStorageRequest {
  int32 shares = 1;
  int32 threshold = 2;
}

message InitStorageResponse {
  repeated string shares = 1;
}

message UnsealRequest {
  repeated string shares = 2;
}

message SealRequest {}

message PutSecretRequest {
  repeated string path = 1;
  string json_value = 2;
  google.protobuf.Timestamp expires_at = 3;
  google.protobuf.Struct metadata = 4;
  int64 if_version = 5;
}

message GetSecretRequest {
  repeated string path = 1;
}

message GetSecretResponse {
  string json_value = 1;
  google.protobuf.Timestamp created_at = 2;
  google.protobuf.Timestamp updated_at = 3;
  google.protobuf.Timestamp expires_at = 4;
  int64 version = 5;
  google.protobuf.Struct metadata = 6;
}

message DeleteSecretRequest {
  repeated string path = 1;
  int64 if_version = 3;
}